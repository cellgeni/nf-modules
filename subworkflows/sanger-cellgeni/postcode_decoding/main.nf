include { IMAGING_POSTCODE } from '../../../modules/sanger-cellgeni/imaging/postcode/main'
include { IMAGING_POSTCODEPREP } from '../../../modules/sanger-cellgeni/imaging/postcodeprep/main'
include { IMAGING_POSTCODEPOST } from '../../../modules/sanger-cellgeni/imaging/postcodepost/main'

workflow POSTCODE_DECODING {
    take:
    ch_profile // channel: [ val(meta), [ profile ] ]
    ch_tabular_codebook // channel: [ val(meta), [ tabular_codebook ] ]
    ch_readout_file // channel: [ val(meta), [ readout_file ] ]
    ch_R // channel: [ val(meta), [ R ] ]
    ch_loc // channel: [ val(meta), [ spot_loc ] ]

    main:

    ch_versions = Channel.empty()

    for_decoding = ch_profile
        .combine(
            ch_tabular_codebook,
            by: 0
        )
        .combine(
            ch_readout_file,
            by: 0
        )
        .combine(ch_R, by: 0)

    // for_decoding.view()

    IMAGING_POSTCODEPREP(for_decoding)
    ch_versions = ch_versions.mix(IMAGING_POSTCODEPREP.out.versions.first())
    // IMAGING_POSTCODEPREP.out.for_decoding.combine(ch_loc, by: 0).view()

    IMAGING_POSTCODE(IMAGING_POSTCODEPREP.out.for_decoding.combine(ch_loc, by: 0))
    ch_versions = ch_versions.mix(IMAGING_POSTCODEPREP.out.versions.first())

    IMAGING_POSTCODEPOST(IMAGING_POSTCODE.out.model_params_and_losses.combine(ch_loc, by: 0))
    ch_versions = ch_versions.mix(IMAGING_POSTCODEPOST.out.versions.first())

    emit:
    processed_profiles = IMAGING_POSTCODEPREP.out.for_decoding // channel: [ val(meta), [ pixel/cell ] ]
    decoded_profiles = IMAGING_POSTCODEPOST.out.decoded_profile // channel: [ val(meta), [ decoded_profile ] ]
    versions = ch_versions // channel: [ versions.yml ]
}
