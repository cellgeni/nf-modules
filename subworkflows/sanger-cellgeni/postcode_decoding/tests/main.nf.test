nextflow_workflow {

    name "Test Subworkflow POSTCODE_DECODING"
    script "../main.nf"
    workflow "POSTCODE_DECODING"

    tag "subworkflows"
    tag "subworkflows_sangercellgeni"
    tag "subworkflows/postcode_decoding"
    tag "postcode_decoding"
    tag "imaging"
    tag "postcode"


    test("postcode - npy - 1") {

        when {
            params {
                postcodeprep_args = "--coding_col_prefix Readouts* --codebook_targer_col gene "
                postcode_args = '--decoding_params {up_prc_to_remove:99.95}'
            }
            workflow {
                """
                input[0] = channel.from([[
                    [ id:'single' ], // meta map
                    file(params.modules_testdata_base_path + '/MERFISH/merfish_transcript_profile_1.npy', checkIfExists: true),
                ]])
                input[1] = channel.from([[
                    [ id:'single' ],
                    file(params.modules_testdata_base_path + '/MERFISH/codebook.csv', checkIfExists: true),
                ]])
                input[2] = channel.from([[
                    [ id:'single' ],
                    file(params.modules_testdata_base_path + '/MERFISH/readouts.csv', checkIfExists: true),
                ]])
                input[3] = channel.from([[
                    [ id:'single' ], 6
                ]])
                input[4] = channel.from([[
                    [ id:'single' ],
                    file(params.modules_testdata_base_path + 'MERFISH/merfish_transcript_locations_1.csv', checkIfExists: true),
                ]])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

    test("postcode - npy - 5000") {

        when {
            params {
                postcodeprep_args = "--coding_col_prefix Readouts* --codebook_targer_col gene "
                postcode_args = '--decoding_params {up_prc_to_remove:99.95}'
            }
            workflow {
                """
                input[0] = channel.from([[
                    [ id:'single' ], // meta map
                    file(params.modules_testdata_base_path + '/MERFISH/merfish_transcript_profile_5000.npy', checkIfExists: true),
                ]])
                input[1] = channel.from([[
                    [ id:'single' ],
                    file(params.modules_testdata_base_path + '/MERFISH/codebook.csv', checkIfExists: true),
                ]])
                input[2] = channel.from([[
                    [ id:'single' ],
                    file(params.modules_testdata_base_path + '/MERFISH/readouts.csv', checkIfExists: true),
                ]])
                input[3] = channel.from([[
                    [ id:'single' ], 6
                ]])
                input[4] = channel.from([[
                    [ id:'single' ],
                    file(params.modules_testdata_base_path + 'MERFISH/merfish_transcript_locations_5000.csv', checkIfExists: true),
                ]])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }


    test("postcode - npy - 100000") {

        when {
            params {
                postcodeprep_args = "--coding_col_prefix Readouts* --codebook_targer_col gene "
                postcode_args = '--decoding_params {up_prc_to_remove:99.95}'
            }
            workflow {
                """
                input[0] = channel.from([[
                    [ id:'single' ], // meta map
                    file(params.modules_testdata_base_path + '/MERFISH/merfish_transcript_profile_100000.npy', checkIfExists: true),
                ]])
                input[1] = channel.from([[
                    [ id:'single' ],
                    file(params.modules_testdata_base_path + '/MERFISH/codebook.csv', checkIfExists: true),
                ]])
                input[2] = channel.from([[
                    [ id:'single' ],
                    file(params.modules_testdata_base_path + '/MERFISH/readouts.csv', checkIfExists: true),
                ]])
                input[3] = channel.from([[
                    [ id:'single' ], 6
                ]])
                input[4] = channel.from([[
                    [ id:'single' ],
                    file(params.modules_testdata_base_path + 'MERFISH/merfish_transcript_locations_100000.csv', checkIfExists: true),
                ]])
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}
