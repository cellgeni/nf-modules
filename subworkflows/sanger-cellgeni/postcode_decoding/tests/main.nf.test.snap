{
    "postcode - npy - 5000": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "single"
                        },
                        "single_decoded_spots.csv:md5,c3f53ca490b62ec46d85591c4770cc01"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "single"
                        },
                        "single_reformatted_profile.npy:md5,ee3dc5826c11d393aaf813b08026c5c2",
                        "single_starfish_codebook.json:md5,846c498cb0b5363e5fcfa44e94bf60b7",
                        "single_barcodes_0123_str.txt:md5,1240bdd581957bc56a810734a5b39b0e"
                    ]
                ],
                "2": [
                    "versions.yml:md5,c105acc99513c405a45c158c81c3cbbc",
                    "versions.yml:md5,c105acc99513c405a45c158c81c3cbbc"
                ],
                "bam": [
                    [
                        {
                            "id": "single"
                        },
                        "single_decoded_spots.csv:md5,c3f53ca490b62ec46d85591c4770cc01"
                    ]
                ],
                "csi": [
                    [
                        {
                            "id": "single"
                        },
                        "single_reformatted_profile.npy:md5,ee3dc5826c11d393aaf813b08026c5c2",
                        "single_starfish_codebook.json:md5,846c498cb0b5363e5fcfa44e94bf60b7",
                        "single_barcodes_0123_str.txt:md5,1240bdd581957bc56a810734a5b39b0e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c105acc99513c405a45c158c81c3cbbc",
                    "versions.yml:md5,c105acc99513c405a45c158c81c3cbbc"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-07-16T15:43:14.418623138"
    },
    "postcode - npy - 100000": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "10000"
                        },
                        "10000_reformatted_profile.npy:md5,45f2e67209caf00d01113dc3802c8519",
                        "10000_starfish_codebook.json:md5,846c498cb0b5363e5fcfa44e94bf60b7"
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "10000"
                        },
                        "10000_decoded.csv:md5,4df11830a543f5c7557aee5807fa93ad"
                    ]
                ],
                "2": [
                    "versions.yml:md5,23c767619679fc9c91e2fe47441281f9",
                    "versions.yml:md5,c105acc99513c405a45c158c81c3cbbc",
                    "versions.yml:md5,c105acc99513c405a45c158c81c3cbbc"
                ],
                "decoded_profiles": [
                    [
                        {
                            "id": "10000"
                        },
                        "10000_decoded.csv:md5,4df11830a543f5c7557aee5807fa93ad"
                    ]
                ],
                "processed_profiles": [
                    [
                        {
                            "id": "10000"
                        },
                        "10000_reformatted_profile.npy:md5,45f2e67209caf00d01113dc3802c8519",
                        "10000_starfish_codebook.json:md5,846c498cb0b5363e5fcfa44e94bf60b7"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,23c767619679fc9c91e2fe47441281f9",
                    "versions.yml:md5,c105acc99513c405a45c158c81c3cbbc",
                    "versions.yml:md5,c105acc99513c405a45c158c81c3cbbc"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-08-05T14:44:26.51547939"
    },
    "postcode - npy - 1": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "single"
                        },
                        "single_reformatted_profile.npy:md5,1767337178874628135d325468822e8a",
                        "single_starfish_codebook.json:md5,846c498cb0b5363e5fcfa44e94bf60b7"
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    "versions.yml:md5,c105acc99513c405a45c158c81c3cbbc",
                    "versions.yml:md5,c105acc99513c405a45c158c81c3cbbc"
                ],
                "decoded_profiles": [
                    
                ],
                "processed_profiles": [
                    [
                        {
                            "id": "single"
                        },
                        "single_reformatted_profile.npy:md5,1767337178874628135d325468822e8a",
                        "single_starfish_codebook.json:md5,846c498cb0b5363e5fcfa44e94bf60b7"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,c105acc99513c405a45c158c81c3cbbc",
                    "versions.yml:md5,c105acc99513c405a45c158c81c3cbbc"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "24.10.5"
        },
        "timestamp": "2025-08-05T14:43:56.244198537"
    }
}